File: api/app/Models/Group.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Group extends Model
{
    use SoftDeletes;

    protected $fillable = [
        'name',
        'vietnamese_name',
        'description',
        'sport_type',
        'skill_level',
        'location',
        'city',
        'district',
        'latitude',
        'longitude',
        'schedule',
        'max_members',
        'min_players',
        'current_members',
        'membership_fee',
        'privacy',
        'auto_approve_members',
        'status',
        'avatar',
        'rules',
        'notification_hours_before',
        'default_locations',
        'sport_specific_settings',
        'creator_id'
    ];

    protected $casts = [
        'schedule' => 'array',
        'rules' => 'array',
        'default_locations' => 'array',
        'sport_specific_settings' => 'array',
        'latitude' => 'decimal:8',
        'longitude' => 'decimal:8',
        'membership_fee' => 'decimal:2',
        'max_members' => 'integer',
        'min_players' => 'integer',
        'current_members' => 'integer',
        'notification_hours_before' => 'integer',
        'auto_approve_members' => 'boolean',
    ];

    public function creator(): BelongsTo
    {
        return $this->belongsTo(User::class, 'creator_id');
    }

    public function memberships(): BelongsToMany
    {
        return $this->belongsToMany(User::class, 'group_memberships')
                    ->withPivot(['role', 'status', 'joined_at', 'left_at', 'join_reason', 'leave_reason', 'total_paid', 'attendance_count', 'member_notes'])
                    ->withTimestamps();
    }

    public function activeMembers(): BelongsToMany
    {
        return $this->memberships()->wherePivot('status', 'hoat_dong');
    }

    public function pendingMembers(): BelongsToMany
    {
        return $this->memberships()->wherePivot('status', 'pending');
    }

    public function getSportNameAttribute(): string
    {
        return match($this->sport_type) {
            'badminton' => 'Cầu lông',
            'football' => 'Bóng đá',
            'tennis' => 'Tennis',
            'pickleball' => 'Pickleball',
            default => $this->sport_type
        };
    }

    public function getSkillLevelNameAttribute(): string
    {
        return match($this->skill_level) {
            'moi_bat_dau' => 'Mới bắt đầu',
            'trung_binh' => 'Trung bình',
            'gioi' => 'Giỏi',
            'chuyen_nghiep' => 'Chuyên nghiệp',
            default => $this->skill_level
        };
    }

    public function getPrivacyNameAttribute(): string
    {
        return match($this->privacy) {
            'cong_khai' => 'Công khai',
            'rieng_tu' => 'Riêng tư',
            default => $this->privacy
        };
    }

    public function getStatusNameAttribute(): string
    {
        return match($this->status) {
            'hoat_dong' => 'Hoạt động',
            'tam_dung' => 'Tạm dừng',
            'dong_cua' => 'Đóng cửa',
            default => $this->status
        };
    }

    public function getDefaultSettings(): array
    {
        return match($this->sport_type) {
            'badminton' => [
                'min_players' => 2,
                'max_players' => 4,
                'notification_hours' => 24,
                'typical_locations' => ['Sân cầu lông', 'Trung tâm thể thao']
            ],
            'football' => [
                'min_players' => 10,
                'max_players' => 22,
                'notification_hours' => 48,
                'typical_locations' => ['Sân bóng đá', 'Sân cỏ nhân tạo']
            ],
            'pickleball' => [
                'min_players' => 2,
                'max_players' => 4,
                'notification_hours' => 12,
                'typical_locations' => ['Sân pickleball', 'Sân tennis đa năng']
            ],
            'tennis' => [
                'min_players' => 2,
                'max_players' => 4,
                'notification_hours' => 12,
                'typical_locations' => ['Sân tennis', 'Câu lạc bộ tennis']
            ],
            default => [
                'min_players' => 2,
                'notification_hours' => 24,
                'typical_locations' => []
            ]
        ];
    }

    public function getAdmins(): BelongsToMany
    {
        return $this->activeMembers()->wherePivot('role', 'admin');
    }

    public function getModerators(): BelongsToMany
    {
        return $this->activeMembers()->wherePivot('role', 'moderator');
    }

    public function isAdmin(User $user): bool
    {
        return $this->memberships()
                    ->where('user_id', $user->id)
                    ->wherePivot('role', 'admin')
                    ->wherePivot('status', 'hoat_dong')
                    ->exists();
    }

    public function isModerator(User $user): bool
    {
        return $this->memberships()
                    ->where('user_id', $user->id)
                    ->wherePivot('role', 'moderator')
                    ->wherePivot('status', 'hoat_dong')
                    ->exists();
    }

    public function isMember(User $user): bool
    {
        return $this->memberships()
                    ->where('user_id', $user->id)
                    ->wherePivot('status', 'hoat_dong')
                    ->exists();
    }

    public function canManage(User $user): bool
    {
        return $user->id === $this->creator_id || $this->isAdmin($user) || $this->isModerator($user);
    }

    /**
     * Get validation rules for group creation
     */
    public static function getValidationRules(): array
    {
        return [
            'name' => 'required|string|max:255',
            'vietnamese_name' => 'nullable|string|max:255',
            'description' => 'nullable|string|max:1000',
            'sport_type' => 'required|in:football,badminton,tennis,pickleball',
            'skill_level' => 'required|in:moi_bat_dau,trung_binh,gioi,chuyen_nghiep',
            'location' => 'required|string|max:500',
            'city' => 'required|string|max:100',
            'district' => 'nullable|string|max:100',
            'latitude' => 'nullable|numeric|between:-90,90',
            'longitude' => 'nullable|numeric|between:-180,180',
            'schedule' => 'nullable|array',
            'max_members' => 'required|integer|min:2|max:50',
            'min_players' => 'nullable|integer|min:1',
            'membership_fee' => 'nullable|numeric|min:0',
            'privacy' => 'required|in:cong_khai,rieng_tu',
            'auto_approve_members' => 'nullable|boolean',
            'status' => 'in:hoat_dong,tam_dung,dong_cua',
            'rules' => 'nullable|array',
            'notification_hours_before' => 'nullable|integer|min:1|max:168',
            'default_locations' => 'nullable|array',
            'sport_specific_settings' => 'nullable|array',
        ];
    }

    /**
     * Get sport-specific validation rules
     */
    public static function getSportSpecificValidationRules(string $sportType): array
    {
        $baseRules = self::getValidationRules();
        
        return match($sportType) {
            'football' => array_merge($baseRules, [
                'min_players' => 'nullable|integer|min:6|max:11',
                'max_members' => 'required|integer|min:10|max:30',
            ]),
            'badminton' => array_merge($baseRules, [
                'min_players' => 'nullable|integer|min:2|max:2',
                'max_members' => 'required|integer|min:2|max:8',
            ]),
            'tennis' => array_merge($baseRules, [
                'min_players' => 'nullable|integer|min:2|max:2',
                'max_members' => 'required|integer|min:2|max:6',
            ]),
            'pickleball' => array_merge($baseRules, [
                'min_players' => 'nullable|integer|min:2|max:2',
                'max_members' => 'required|integer|min:2|max:8',
            ]),
            default => $baseRules
        };
    }

    /**
     * Validate Vietnamese name pattern
     */
    public static function isValidVietnameseName(string $name): bool
    {
        // Allow Vietnamese characters, numbers, spaces, and common punctuation
        return preg_match('/^[\p{L}\p{N}\s\-_.()]+$/u', $name) === 1;
    }
}

File: mobile/lib/features/groups/models/group.dart
// Hot reload trigger - fix navigation
class Group {
  final int id;
  final String name;
  final String? description;
  final String sportType;
  final String skillLevel;
  final String location;
  final String city;
  final String? district;
  final double? latitude;
  final double? longitude;
  final Map<String, dynamic>? schedule;
  final int maxMembers;
  final int currentMembers;
  final double membershipFee;
  final String privacy;
  final String status;
  final String? avatar;
  final Map<String, dynamic>? rules;
  final int creatorId;
  final DateTime createdAt;
  final DateTime updatedAt;
  final User? creator;
  final List<User>? activeMembers;
  final List<GroupMember>? members;

  Group({
    required this.id,
    required this.name,
    this.description,
    required this.sportType,
    required this.skillLevel,
    required this.location,
    required this.city,
    this.district,
    this.latitude,
    this.longitude,
    this.schedule,
    required this.maxMembers,
    required this.currentMembers,
    required this.membershipFee,
    required this.privacy,
    required this.status,
    this.avatar,
    this.rules,
    required this.creatorId,
    required this.createdAt,
    required this.updatedAt,
    this.creator,
    this.activeMembers,
    this.members,
  });

  factory Group.fromJson(Map<String, dynamic> json) {
    return Group(
      id: json['id'] as int? ?? 0,
      name: json['name'] as String? ?? 'Chưa có tên',
      description: json['description'] as String?,
      sportType: json['sport_type'] as String? ?? 'unknown',
      skillLevel: json['skill_level'] as String? ?? 'moi_bat_dau',
      location: json['location'] as String? ?? 'Chưa xác định',
      city: json['city'] as String? ?? 'Chưa xác định',
      district: json['district'] as String?,
      latitude: _convertToDouble(json['latitude']),
      longitude: _convertToDouble(json['longitude']),
      schedule: json['schedule'] is List 
          ? (json['schedule'] as List).isEmpty ? null : null
          : json['schedule'] as Map<String, dynamic>?,
      maxMembers: json['max_members'] as int? ?? 20,
      currentMembers: json['current_members'] as int? ?? 0,
      membershipFee: _convertToDouble(json['membership_fee'] ?? json['monthly_fee']) ?? 0.0,
      privacy: json['privacy'] as String? ?? 'cong_khai',
      status: json['status'] as String? ?? 'hoat_dong',
      avatar: json['avatar'] as String?,
      rules: json['rules'] is List 
          ? (json['rules'] as List).isEmpty ? null : null
          : json['rules'] as Map<String, dynamic>?,
      creatorId: json['creator_id'] as int? ?? 0,
      createdAt: json['created_at'] != null 
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : DateTime.now(),
      creator: json['creator'] != null ? User.fromJson(json['creator']) : null,
      activeMembers: json['active_members'] != null
          ? (json['active_members'] as List)
              .map((member) => User.fromJson(member))
              .toList()
          : null,
      members: json['members'] != null
          ? (json['members'] as List)
              .map((member) =>
                  GroupMember.fromJson(member as Map<String, dynamic>))
              .toList()
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'sport_type': sportType,
      'skill_level': skillLevel,
      'location': location,
      'city': city,
      'district': district,
      'latitude': latitude,
      'longitude': longitude,
      'schedule': schedule,
      'max_members': maxMembers,
      'current_members': currentMembers,
      'membership_fee': membershipFee,
      'privacy': privacy,
      'status': status,
      'avatar': avatar,
      'rules': rules,
      'creator_id': creatorId,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  Group copyWith({
    int? id,
    String? name,
    String? description,
    String? sportType,
    String? skillLevel,
    String? location,
    String? city,
    String? district,
    double? latitude,
    double? longitude,
    Map<String, dynamic>? schedule,
    int? maxMembers,
    int? currentMembers,
    double? membershipFee,
    String? privacy,
    String? status,
    String? avatar,
    Map<String, dynamic>? rules,
    int? creatorId,
    DateTime? createdAt,
    DateTime? updatedAt,
    User? creator,
    List<User>? activeMembers,
    List<GroupMember>? members,
  }) {
    return Group(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      sportType: sportType ?? this.sportType,
      skillLevel: skillLevel ?? this.skillLevel,
      location: location ?? this.location,
      city: city ?? this.city,
      district: district ?? this.district,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      schedule: schedule ?? this.schedule,
      maxMembers: maxMembers ?? this.maxMembers,
      currentMembers: currentMembers ?? this.currentMembers,
      membershipFee: membershipFee ?? this.membershipFee,
      privacy: privacy ?? this.privacy,
      status: status ?? this.status,
      avatar: avatar ?? this.avatar,
      rules: rules ?? this.rules,
      creatorId: creatorId ?? this.creatorId,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      creator: creator ?? this.creator,
      activeMembers: activeMembers ?? this.activeMembers,
      members: members ?? this.members,
    );
  }

  String get sportName {
    switch (sportType) {
      case 'cau_long':
        return 'Cầu lông';
      case 'bong_da':
        return 'Bóng đá';
      case 'bong_ro':
        return 'Bóng rổ';
      case 'tennis':
        return 'Tennis';
      case 'bong_chuyen':
        return 'Bóng chuyền';
      case 'bong_ban':
        return 'Bóng bàn';
      case 'chay_bo':
        return 'Chạy bộ';
      case 'dap_xe':
        return 'Đạp xe';
      case 'boi_loi':
        return 'Bơi lội';
      case 'yoga':
        return 'Yoga';
      case 'gym':
        return 'Gym';
      default:
        return sportType;
    }
  }

  String get skillLevelName {
    switch (skillLevel) {
      case 'moi_bat_dau':
        return 'Mới bắt đầu';
      case 'trung_binh':
        return 'Trung bình';
      case 'gioi':
        return 'Giỏi';
      case 'chuyen_nghiep':
        return 'Chuyên nghiệp';
      default:
        return skillLevel;
    }
  }

  String get privacyName {
    switch (privacy) {
      case 'cong_khai':
        return 'Công khai';
      case 'rieng_tu':
        return 'Riêng tư';
      default:
        return privacy;
    }
  }

  String get statusName {
    switch (status) {
      case 'hoat_dong':
        return 'Hoạt động';
      case 'tam_dung':
        return 'Tạm dừng';
      case 'dong_cua':
        return 'Đóng cửa';
      default:
        return status;
    }
  }

  bool get isFull => currentMembers >= maxMembers;
  bool get isActive => status == 'hoat_dong';
  bool get isPublic => privacy == 'cong_khai';

  static double? _convertToDouble(dynamic value) {
    if (value == null) return null;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) {
      return double.tryParse(value);
    }
    return null;
  }
}

class User {
  final int id;
  final String name;
  final String phone;
  final String? avatar;

  User({
    required this.id,
    required this.name,
    required this.phone,
    this.avatar,
  });

  factory User.fromJson(Map<String, dynamic> json) {
    return User(
      id: json['id'] as int? ?? 0,
      name: json['name'] as String? ?? 'Người dùng',
      phone: json['phone'] as String? ?? '',
      avatar: json['avatar'] as String?,
    );
  }
}

class GroupMember {
  final int userId;
  final String name;
  final String role;
  final String? avatar;
  final bool isCurrentUser;

  GroupMember({
    required this.userId,
    required this.name,
    required this.role,
    this.avatar,
    this.isCurrentUser = false,
  });

  factory GroupMember.fromJson(Map<String, dynamic> json) {
    return GroupMember(
      userId: (json['user_id'] ?? json['id'] ?? 0) as int,
      name: json['name'] as String? ?? 'Thành viên',
      role: (json['role'] ?? json['group_role'] ?? 'member') as String,
      avatar: json['avatar'] as String?,
      isCurrentUser: json['is_current_user'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'name': name,
      'role': role,
      'avatar': avatar,
      'is_current_user': isCurrentUser,
    };
  }
}
